Simplify and avoid redundant typecasts in build_object_descriptor

smowton opened the pull request:
Otherwise this produces odd expressions like (1 + (int)2), which can lead to missed opportunities since that doesn't pass `is_constant()`

smowton committed:
build_object_descriptor: avoid redundant typecasts

These can fool later simplifications by appearing to be non-constant.

smowton committed:
Simplify object descriptor offset after construction

Leaving this with trivially simplifiable expressions such as (0 + 0) means subsequent simplifications that
check for offset.is_constant() may be wrongly skipped.

tautschnig commented on src/util/std_expr.cpp:
```diff
@@ -145,6 +146,7 @@ void object_descriptor_exprt::build(
     offset()=from_integer(0, index_type());
 
   build_object_descriptor_rec(ns, expr, *this);
+  simplify(offset(), ns);
```
I'm a bit worried about the cost, can you check that callees don't currently run this after invoking build? (And if they do, remove it from those places.)

smowton replied:
Checked, there was one redundant simplify (presumably working around the same issue). A couple of others use the offset to build a composite expression then simplify *that*, and so would have coincidentally dodged this problem. Some others uses `ode.offset().is_zero()`, which will probably have failed spuriously before when the offset was composite-but-constant.

smowton committed:
Generics unit test: set integer widths

This test indirectly uses object_descriptor_exprt, which uses simplify_sum, which in turn requires
integer constants to have defined widths.

smowton committed:
Remove now-unnecessary simplify after object_descriptor_exprt::build

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 8c231e96deda90fff2b35db505f626bb4042a2fa).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100903764

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7d0ce455edb18026439f509a70f4fdf8673f4ecd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100908538
